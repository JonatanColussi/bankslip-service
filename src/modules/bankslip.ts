import { BANKSLIP_TYPES } from '../constants';
import { BankslipTypes } from '../types';

export class BankslipBase {
  amount?: number;
  barCode?: string;
  expirationDate?: string;
  type?: BankslipTypes;

  constructor(protected readonly digitableLine: string) {
    this.validateDigitableLine();
    this.setBankslipType();
  }

  protected formatAmount(amount: string) {
    const integer = amount.substring(0, amount.length - 2);
    const cents = amount.slice(-2);

    this.amount = Number.parseFloat(`${integer}.${cents}`) || undefined;
  }

  private setBankslipType() {
    this.type = this.digitableLine.startsWith('8')
      ? BANKSLIP_TYPES.CONVENIO
      : BANKSLIP_TYPES.TITULO;
  }

  private validateDigitableLine() {
    if (!/^\d{47,48}$/.test(this.digitableLine)) {
      throw new Error('The digitable line must contains only numbers');
    }
  }
}
