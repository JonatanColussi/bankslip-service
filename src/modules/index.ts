export * from './bankslip';
export * from './convenio';
export * from './titulo';
