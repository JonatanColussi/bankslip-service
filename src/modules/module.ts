export function module10(block: string) {
  const segments = [...block].map(n => Number(n)).reverse();

  const totalSum = segments.reduce((acc, current, index) => {
    let sum = current * (((index + 1) % 2) + 1);

    sum = sum > 9 ? Math.trunc(sum / 10) + (sum % 10) : sum;

    return acc + sum;
  }, 0);

  return Math.ceil(totalSum / 10) * 10 - totalSum;
}

export function module11(block: string) {
  const segments = [...block].map(n => Number(n)).reverse();

  let multiplier = 2;
  const totalSum = segments.reduce((acc, current) => {
    const sum = current * multiplier;

    multiplier = multiplier === 9 ? 2 : multiplier + 1;

    return acc + sum;
  }, 0);

  const rest = totalSum % 11;

  if ([0, 1].includes(rest)) {
    return 0;
  }

  if (rest === 10) {
    return 1;
  }

  const DV = 11 - rest;

  return DV;
}
