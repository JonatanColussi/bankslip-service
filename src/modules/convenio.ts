import { BankslipBase } from './bankslip';
import { module10, module11 } from './module';

import { BANKSLIP_TYPES } from '../constants';
import { Bankslip } from '../types';

export class Convenio extends BankslipBase implements Bankslip {
  constructor(protected readonly digitableLine: string) {
    super(digitableLine);
    this.validate();
    this.setAmount();
    this.setBarcode();
    this.setExpirationDate();
  }

  setAmount() {
    const reference = Number(this.digitableLine.substring(2, 3));
    const hasAmount = [6, 8].includes(reference);

    if (!hasAmount) {
      this.amount = undefined;

      return;
    }

    const amount = [...this.digitableLine];

    amount.splice(11, 1);

    this.formatAmount(amount.join('').substring(4, 15));
  }

  setBarcode() {
    const barcode = [...this.digitableLine];

    barcode.splice(11, 1);
    barcode.splice(22, 1);
    barcode.splice(33, 1);
    barcode.splice(44, 1);

    this.barCode = barcode.join('');
  }

  setExpirationDate() {
    this.expirationDate = undefined;
  }

  validate() {
    if (this.type !== BANKSLIP_TYPES.CONVENIO) {
      throw new Error('invalid instance type');
    }

    const reference = Number(this.digitableLine.substring(2, 3));
    const module = [6, 7].includes(reference) ? module10 : module11;

    const blocks = [
      {
        block: this.digitableLine.substring(0, 11),
        dv: this.digitableLine.substring(11, 12),
      },
      {
        block: this.digitableLine.substring(12, 23),
        dv: this.digitableLine.substring(23, 24),
      },
      {
        block: this.digitableLine.substring(24, 35),
        dv: this.digitableLine.substring(35, 36),
      },
      {
        block: this.digitableLine.substring(36, 47),
        dv: this.digitableLine.substring(47, 48),
      },
    ];

    if (!blocks.every(item => module(item.block) === Number(item.dv))) {
      throw new Error('The digitable line has invalid checksum');
    }
  }
}
