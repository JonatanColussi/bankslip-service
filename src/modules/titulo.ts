import { BankslipBase } from './bankslip';
import { module10 } from './module';

import { BANKSLIP_TYPES } from '../constants';
import { Bankslip } from '../types';

export class Titulo extends BankslipBase implements Bankslip {
  constructor(protected readonly digitableLine: string) {
    super(digitableLine);
    this.validate();
    this.setAmount();
    this.setBarcode();
    this.setExpirationDate();
  }

  setAmount() {
    const amount = this.digitableLine.substring(37);

    this.formatAmount(amount);
  }

  setBarcode() {
    this.barCode =
      this.digitableLine.substring(0, 4) +
      this.digitableLine.substring(32, 47) +
      this.digitableLine.substring(4, 9) +
      this.digitableLine.substring(10, 20) +
      this.digitableLine.substring(21, 31);
  }

  setExpirationDate() {
    const date = new Date(Date.UTC(1997, 9, 7));
    const days = Number.parseInt(this.digitableLine.substring(33, 37), 10);

    if (Number.isNaN(days) || days === 0) {
      this.expirationDate = undefined;

      return;
    }

    date.setDate(date.getDate() + days);

    const [expirationDate] = date.toISOString().split('T');

    this.expirationDate = expirationDate;
  }

  validate() {
    if (this.type !== BANKSLIP_TYPES.TITULO) {
      throw new Error('invalid instance type');
    }

    const blocks = [
      {
        block: this.digitableLine.substring(0, 9),
        dv: this.digitableLine.substring(9, 10),
      },
      {
        block: this.digitableLine.substring(10, 20),
        dv: this.digitableLine.substring(20, 21),
      },
      {
        block: this.digitableLine.substring(21, 31),
        dv: this.digitableLine.substring(31, 32),
      },
    ];

    if (!blocks.every(item => module10(item.block) === Number(item.dv))) {
      throw new Error('The digitable line has invalid checksum');
    }
  }
}
