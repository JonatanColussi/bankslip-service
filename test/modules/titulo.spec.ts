import { Titulo } from 'modules/titulo';

it('should return an error if intance with invalid bankslip type', () => {
  expect(() => new Titulo('836100000055483500863343933736760093100258416914')).toThrow(
    'invalid instance type',
  );
});

it('should return an error if checksum is invalid', () => {
  expect(() => new Titulo('23793381286000795713695000063305475520000370000')).toThrow(
    'The digitable line has invalid checksum',
  );
});

it.each([
  '33690000090000001004100731323143190290002085742',
  '03399492813698200090709545001027500000000000000',
  '00190000090250038600400684863178390220000048550',
  '21290001192110001210904475617405975870000002000',
])('should parse bankslip properly', digitableLine => {
  expect(new Titulo(digitableLine)).toMatchSnapshot();
});
