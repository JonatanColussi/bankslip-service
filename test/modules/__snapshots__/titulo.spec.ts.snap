// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse bankslip properly 1`] = `
Titulo {
  "amount": 20857.42,
  "barCode": "33691902900020857420000000000010040073132314",
  "digitableLine": "33690000090000001004100731323143190290002085742",
  "expirationDate": "2022-06-27",
  "type": "TITULO",
}
`;

exports[`should parse bankslip properly 2`] = `
Titulo {
  "amount": undefined,
  "barCode": "03395000000000000009492836982000900954500102",
  "digitableLine": "03399492813698200090709545001027500000000000000",
  "expirationDate": undefined,
  "type": "TITULO",
}
`;

exports[`should parse bankslip properly 3`] = `
Titulo {
  "amount": 485.5,
  "barCode": "00193902200000485500000002500386000068486317",
  "digitableLine": "00190000090250038600400684863178390220000048550",
  "expirationDate": "2022-06-20",
  "type": "TITULO",
}
`;

exports[`should parse bankslip properly 4`] = `
Titulo {
  "amount": 20,
  "barCode": "21299758700000020000001121100012100447561740",
  "digitableLine": "21290001192110001210904475617405975870000002000",
  "expirationDate": "2018-07-16",
  "type": "TITULO",
}
`;
